### PR TITLE
[Feature] Make insight annotation a vote instead of single user decision

### PR DESCRIPTION
**Pull request by zond**
- New AnnotationVote data model.
- Posting answers (when annotating insights) now records username and device ID into a vote instead of annotating the insight.
  - If enough votes are present (3 votes for any option) then a decision is made.
  - If no other annotation has >1 vote, the annotation with 3 votes is saved into the insight, otherwise decision 0 (don't know) is saved.
- Loading questions (when annotating insights) now prioritizes voted-on insights (by sorting them partly by number of votes) to reduce time to get partly annotated insights finished.
- Loading questions (when annotating insights) now filters out insights the user or device have voted on previously.
- New test target for integration tests requiring the entire dev environment to test the endpoint changes.

### CODE REVIEW

File: .env
Comment (zond) [inline .env:15]: This is _only_ used (AFAIU) to download a JSON file from PO.

I'm not very happy about it - what alternatives do we have?

File: Makefile
Comment (zond) [inline Makefile:75]: I added this just to get my `make` command when first checking out the code to work. Has anyone else experienced that problem? This doesn't feel like the correct solution.

File: docker/dev.yml
Comment (zond) [inline docker/dev.yml:29]: We might want to remove this, or does anyone think it might be useful in the future?

**Comment by zond** [comment]
It annoys me a lot that we have GitHub actions that fail due to things outside this PR. How prioritized is it to get those things working?

File: .env
Comment (kulizhsy) [reply .env:15]: Not only for JSON - look at the usage of `BaseURLProvider`.

What are you not happy about in particular, the file downloads/the fact we have to manipulate URLs to get things working?

File: docker/dev.yml
Comment (kulizhsy) [reply docker/dev.yml:29]: Since we can access the db through docker, I'm not sure if we also want to expose the ports

File: .env
Comment (zond) [reply .env:15]: Hm, maybe I should have clarified - I meant in the integration tests. But it might not be true even there of course :)

What I'm unhappy with is how the test isn't hermetic - if the .net environment is down the tests will fail.

File: docker/dev.yml
Comment (zond) [reply docker/dev.yml:29]: Alright, I'll remove it.

File: .env
Comment (kulizhsy) [reply .env:15]: We can construct a test taxonomy JSON with the one category we need in tests and mock out the actual call, returning the test taxonomy instead?

Comment (kulizhsy) [reply .env:15]: Actually, seems we can just use a [similar set up to this test](https://github.com/openfoodfacts/robotoff/blob/837c2da66b80e48a47788dc92d7975bd6c569424/tests/test_taxonomy.py#L9), since Robotoff already has a fallback taxonomy JSON stored

Comment (zond) [reply .env:15]: That's a promising idea, but the test you linked doesn't make the API in general use a different source for JSON files, it just validates that using those local files works, right?

I'm looking at how to tell the entire Robotoff engine to use the local files instead..

Comment (zond) [reply .env:15]: That's a promising idea, but the test you linked doesn't make the API in general use a different source for JSON files, it just validates that using those local files works, right?

I'm looking at how to tell the entire Robotoff engine to use the local files instead..

Comment (zond) [reply .env:15]: Alright, when I started looking (and trying to make the taxonomy JSON specifically load locally) I noticed that Robotoff uses PO to load product images and things as well.

I think it will have to rely on the .net environment until we get enough problems from it that it's worth either mocking up a productopener, or forcing the integration tests to run a local productopener as well.

File: docker/dev.yml
Comment (ocervell) [reply docker/dev.yml]: +1

**Comment by zond** [comment]
I tried to do it the standard git way, using `git rebase master`, but it lead to horrible and repeated merge problems that just didn't happen when I tried a simple `git merge master` instead. I'm OK with `merge` instead of `rebase` (I prefer it even), but anyone complaining can please explain to me what I did wrong first :)

File: robotoff/app/core.py
Comment (alexgarel) [inline robotoff/app/core.py:102]: I'm curious if this is effective enough on a large number of insight, to make voted one really get first places ? (I can see it multiplies expected value, but I have the feeling that on a large collection, we won't get so much voted results !)

File: robotoff/app/core.py
Comment (alexgarel) [inline robotoff/app/core.py:123]: Opinion: I find the name a bit misleading. maybe verified_annotation (True meaning we do not use voting) could be more explicit. Or trusted_annontation.

File: robotoff/app/core.py
Comment (alexgarel) [inline robotoff/app/core.py:186]: we do not seem to have an enum for this, a pity !

File: robotoff/app/core.py
Comment (alexgarel) [inline robotoff/app/core.py:151]: why don't we run all the save_annotation function atomically ? (that would make more sense).

Comment (alexgarel) [reply robotoff/app/core.py:151]: in fact in annotator.annotate we will open a distinct transaction. But doesn't that mean, we can be in an inconsistent state ?

File: robotoff/app/core.py
Comment (alexgarel) [inline robotoff/app/core.py:142]: Opinion: that would be better to have a separate method for this part.

File: robotoff/app/middleware.py
Comment (alexgarel) [inline robotoff/app/middleware.py:6]: Nice.

**Review by alexgarel — COMMENTED** [review]
I didn't review the tests yet, but prefer to send my comments so far.

I'm rather picky in my comments, so feel free to ignore some of them.

Really cool to have distinguished integration tests from unit tests.

File: robotoff/models.py
Comment (alexgarel) [reply robotoff/models.py:80]: I've seen why in the code (we wan't to sort on it), so documenting is enough.

File: Makefile
Comment (zond) [reply Makefile:75]: Should I have done that manually? Is there no way to do it automatically?

File: robotoff/models.py
Comment (zond) [reply robotoff/models.py:156]: I tried to not even provide `id`, but it didn't work - and since I'm more used to autoincrementing ints for PKs, I thought this was the way you do it with UUIDs. What is the correct way?

File: robotoff/models.py
Comment (zond) [reply robotoff/models.py:164]: Added `choices`.

File: robotoff/models.py
Comment (zond) [reply robotoff/models.py:80]: Added documenting comment.

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py:79]: Mostly because it's only used inside this particular function. If you feel strongly about I'll change it, but it works with my taste ;)

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py:89]: It would actually work with both, but it would be a potentially more expensive JOIN. We could of course enforce that it's not used that way, but it didn't feel important. WDYT?

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py:102]: If I'm not mistaken, insights with n+1 votes will be (n+1)/(n) more likely to bubble to the top than insights with n votes. It's not perfect, but it doesn't need to be. I suggest we wait and see how it ends up working?

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py:123]: I also found it a bit confusing, but was happy with the comment in the beginning of `save_annotation`. I defer to @kulizhsy for this discussion.

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py:186]: Is this you wanting us to change something in the PR, or is it just a sentiment about the depressing state of the world? :)

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py:151]: The code would be more complex if we wanted all of it in a transaction, and transactions are expensive.

At the same time, the only thing important to keep atomic there is that the n_votes is correct, so that insights don't stay at the bottom of the pile by mistake.

If two votes arrive at the same time, the worst thing that could happen currently is that the second vote would override the first vote, but I'm not sure that is bad enough to warrant extra code complexity.

WDYT?

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py:142]: I think that's a part of a reasonable cleanup with the transactions as well. A proper refactoring would probably split it out into one function that made votes and optionally also called a function that annotated insights if enough votes were collected - and the second function would either reuse or create a new transaction as needed.

Do we need it for this PR though?

File: robotoff/app/middleware.py
Comment (zond) [reply robotoff/app/middleware.py:6]: Yeah, it was necessary to be able to run the integration tests. Not sure why - but it seems the app starts up twice when I run them..?

File: robotoff/models.py
Comment (kulizhsy) [reply robotoff/models.py:157]: I'm not sure if it's a peewee issue or a postgres issue, but there indeed does not seem to be a way to just leave it unset, without a default, and have an ID auto-assigned.

I saw an alternative way to have it auto-assigned with   `peewee.UUIDField(.. ,constraints=[SQL("DEFAULT uuid_generate_v4()")])`

but that also was throwing errors for me..

File: robotoff/app/core.py
Comment (kulizhsy) [reply robotoff/app/core.py:123]: I like Alex's suggestion of the name `verified_annotation` & flipping of the default.

Comment (zond) [reply robotoff/app/core.py:123]: @kulizhsy will you change it or should I?

File: robotoff/models.py
Comment (zond) [reply robotoff/models.py:157]: Let's leave it like this then. Feel free to reopen the conversation if you disagree :D

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py:89]: @kulizhsy does that mean you volunteer to implement it? :)

Comment (kulizhsy) [reply robotoff/app/core.py:89]: Sure, I'll do it :)

File: robotoff/app/core.py
Comment (kulizhsy) [reply robotoff/app/core.py:123]: Sure, let's go with `trusted_annotator`, I'll update the code.

File: robotoff/app/core.py
Comment (alexgarel) [reply robotoff/app/core.py:118]: yep, it's just that if you have 1000 insight, and 10 with 2 votes, you will only have  20 / 1010 chances of getting an already voted one. But You're right that this has the advantage of being very simple :-)
Otherwise I was thinking something like:
random / ((4 - votes) * alpha)
where alpha could be 4

File: robotoff/app/core.py
Comment (alexgarel) [reply robotoff/app/core.py:167]: I'm ok with it.
As you say there no really evil consequences.

File: .github/workflows/pull_request.yml
Comment (kulizhsy) [inline .github/workflows/pull_request.yml:104]: This seems to break the github action:

`Service 'workers' failed to build : failed to export image: failed to create image: failed to get layer sha256:b5c7da884f67639fd3dec4ab67e5950371fdfed950da68b3dc0b40f5199d8682: layer does not exist`

[Full log](https://github.com/openfoodfacts/robotoff/runs/4093308717?check_suite_focus=true)

@alexgarel @zond how do we fix this?

File: robotoff/app/core.py
Comment (alexgarel) [reply robotoff/app/core.py:202]: It was a gentle push about if you feel like fixing it in this PR (if you were already on the edge of doing it).

File: .github/workflows/pull_request.yml
Comment (zond) [reply .github/workflows/pull_request.yml:104]: Can it be temporary network problems? I looks like the docker image repo was broken.

Comment (kulizhsy) [reply .github/workflows/pull_request.yml:104]: Not sure, but it was[ failing yesterday](https://github.com/openfoodfacts/robotoff/runs/4066110863?check_suite_focus=true) too

Comment (zond) [reply .github/workflows/pull_request.yml:104]: Does it work locally for you? For me they do.

@alexgarel, @ocervell do you know why the tests would fail due to failure to get a docker image - on GitHub but not locally...?

File: tests/integration/test_api.py
Comment (alexgarel) [inline tests/integration/test_api.py:147]: wow, never seen that kind of dict_items comparison before ! That's cool.

File: tests/integration/test_api.py
Comment (alexgarel) [inline tests/integration/test_api.py:184]: you don't need to put list "[" in all, you can directly write:
`all(val is None for key, val in insight.items() if key in {"username", "completed_at", "annotation"})`

I would even have written:
`not any(insight[key] for key in ("username", "completed_at", "annotation"))`

File: tests/integration/test_api.py
Comment (alexgarel) [inline tests/integration/test_api.py:336]: Good test covarage ! :+1:

**Review by alexgarel — APPROVED** [review]
Ok, some more pedantic comments, but overall it's ok !

File: robotoff/app/api.py
Comment (zond) [reply robotoff/app/api.py:842]: Fixed it.

File: Makefile
Comment (zond) [reply Makefile:75]: Super annoying, there should always be a default that just works I feel. But I'll remove the chmod :)

File: tests/integration/test_api.py
Comment (zond) [reply tests/integration/test_api.py:105]: Deferring this to @kulizhsy

File: tests/integration/test_api.py
Comment (zond) [reply tests/integration/test_api.py:147]: What does it actually do? I sort of feel that it should check that it has "at least those elements", but I can't say I feel certain about it.

Comment (kulizhsy) [reply tests/integration/test_api.py:147]: @zond, yep, it does what you said :)

Comment (zond) [reply tests/integration/test_api.py:147]: Nice one, thanks!

Comment (alexgarel) [reply tests/integration/test_api.py:147]: Yes python docs says dict_items acts as a Set of (key, value). And for sets, A > B means that A got all the elements from B and some more.

File: Makefile
Comment (ocervell) [reply Makefile]: I don't think we ***need*** to set `OFF_UID` for Robotoff --> this adds a step that is slowing local dev imo. For PO we do need to change ownership of data folders coming from external, but for Robotoff I don't think it should be necessary.

File: .github/workflows/pull_request.yml
Comment (alexgarel) [reply .github/workflows/pull_request.yml:104]: Seems like a known bug see https://stackoverflow.com/questions/51115856/docker-failed-to-export-image-failed-to-create-image-failed-to-get-layer#62409523 and https://github.com/moby/moby/issues/37965
Some suggest adding a `RUN true` to Dockerfile after `COPY` or using BUILDKIT (which I would like :-) )

@ocervell what do you think ?

Comment (ocervell) [reply .github/workflows/pull_request.yml:104]: Never seen this issue before, we can try using BUILDKIT by setting the env variable `DOCKER_BUILDKIT=1` in the step `env`.

File: Makefile
Comment (alexgarel) [reply Makefile]: I disagree, in the sense, that you will end up with folders on your machine, that you can't list / copy etc. without using sudo, which seems worst to me. (but yes it's a compromise to be found)

For me as soon as you:
- don't use root in your container (security best practice)
- are a local dev (not on machine, where you handle multi user and usually connect with root / sudoer) (this may apply if you admin stuff with non sudoers users on the system)
- bind mount folders in docker

You have to align container user and your id.

Typically here, for the models folders, how do you handle it otherwise, appart from setting `chmod a+x models/` or `sudo chown 1000:1000 models/` ?

File: tests/integration/test_api.py
Comment (kulizhsy) [reply tests/integration/test_api.py:105]: I'll update the tests

File: tests/integration/test_api.py
Comment (kulizhsy) [reply tests/integration/test_api.py:184]: Done, thanks!

File: .github/workflows/pull_request.yml
Comment (kulizhsy) [reply .github/workflows/pull_request.yml:104]: Hmm, `DOCKER_BUILDKIT=1` didn't seem to help :)

File: Makefile
Comment (zond) [reply Makefile]: Git only seems to care about executable permissions unfortunately.

Do we even _need_ OFF_UID for Robotoff? Why not just run everything as current user?

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py]: No more comments received, so I guess I'll resolve this.

Comment (zond) [reply robotoff/app/core.py]: Ah, I see it's also redundant after another change :D

File: robotoff/app/core.py
Comment (zond) [reply robotoff/app/core.py]: And I see it's done. Nice!

**Comment by zond** [comment]
We only have very few unresolved comments now, almost ready to merge to master. However, the automatic testing still doesn't work. In particular `DeepSource` and `Robotoff code quality checks and unit tests`. Does anyone have ideas why?

**Review by stephanegigandet — APPROVED** [review]

File: .github/workflows/pull_request.yml
Comment (alexgarel) [reply .github/workflows/pull_request.yml:104]: I added the `RUN true`, let see if it fails or not…

**Comment by alexgarel** [comment]
@kulizhsy  
1. I lowered requirements in deepsource on documentation, and tell it not to block PRs (let see if it works)
2. I fixed the problem with docker
3. I changed the way integration tests are runned (exec is not the good solution from what I know), and removed the `-v` to `docker down` as I want to run tests locally without removing my volumes !

Now if I run tests locally, I run into problems because of pewee having a failed transaction. I think it should be handled by the pytest fixture : robustly rollback any pending transaction after the script.

File: robotoff/models.py
Comment (alexgarel) [inline robotoff/models.py:82]: (sorry I comment here to have a new thread).

We add a column, but how do we handle database migration ?

Integration tests are failing in my case because my database already exists and does not have this column. Did I missed some step ?

Maybe we should use http://docs.peewee-orm.com/en/latest/peewee/playhouse.html#schema-migrations

File: Makefile
Comment (alexgarel) [inline Makefile:122]: The way we run integrations tests, without changing database name, means we are destructive on the dev desktop. 

I think we can tackle that as a separate issue, but do you agree this is a problem ? (I do find it an important problem)

File: .github/workflows/pull_request.yml
Comment (alexgarel) [reply .github/workflows/pull_request.yml:104]: This does the trick, now let see if it pass or fail for other reasons.

File: robotoff/models.py
Comment (alexgarel) [reply robotoff/models.py:82]: @zond **this is a blocker**

Comment (zond) [reply robotoff/models.py:82]: As I said in the chat - I think we could migrate manually this time. AFAIU this is the first ever DB change in Robotoff, and if migrations only happen every 3 years maybe we don't need to build a migration system this time?

An alternative is to build migration support in a separate PR, and merge that one into this, and add a proper migration.

WDYT?

File: Makefile
Comment (alexgarel) [reply Makefile:127]: I do :-)

File: robotoff/models.py
Comment (zond) [reply robotoff/models.py:82]: IIRC we agreed to do this migration manually? Closing this discussion.

File: Makefile
Comment (zond) [reply Makefile:127]: Alright, let's leave that as topic for a new PR?

Comment (alexgarel) [reply Makefile:127]: ok.

Comment (alexgarel) [reply Makefile:127]: I opened an issue : https://github.com/openfoodfacts/robotoff/issues/490

File: robotoff/models.py
Comment (alexgarel) [inline robotoff/models.py:80]: @zond @kulizhsy here is the SQL I'am planning to run to migrate:

```
ALTER TABLE product_insight ADD COLUMN IF NOT EXISTS n_votes integer NOT NULL DEFAULT 0;
```
I think the best is to run it ahead of deployment, as it should not disturb.

I am wondering, if I should also run a :
```
UPDATE product_insight SET n_votes=1 where annotation = 1;
```

Comment (zond) [reply robotoff/models.py:80]: The migration SQL seems correct.

I don't think you need to run the other code - we'll just have to keep in mind that old annotations are annotated without votes. Setting `n_votes` without actually populating the vote table seems dangerous. Better to keep `n_votes` at zero and use it to identify old annotations without votes?

**Comment by ocervell** [comment]
@alexgarel can we merge this ? Are there opened issues still ?

File: robotoff/models.py
Comment (alexgarel) [reply robotoff/models.py:80]: I've run the migration on both preprod and actual prod (10.1.0.103)

cc: @ocervell

**Comment by alexgarel** [comment]
@ocervell @zond and @kulizhsy 

I would be ready to Merge, now it seems to me, it would be better to have some time to test it in preprod.

And as we are about to deploy to new prod (probably on monday), I wonder if we should wait to merge this one after that ?

**Pull request merged**
